src/osm4routing/reader.rs

```diff
@@ -1,259 +1,434 @@
+              missing_node = true;
+              missing_node_stats.push((way_index, *node_id));
+              break;
+          }
+      }
+
+      if missing_node {
+          ways_to_remove.push(way_index);
+          removed_ways_count += 1;
+      }
     }
 
-    fn split_way(&self, way: &Way) -> Vec<Edge> {
-        let mut result = Vec::new();
-
-        let mut source = NodeId(0);
-        let mut geometry = Vec::new();
-        let mut nodes = Vec::new();
-        for (i, &node_id) in way.nodes.iter().enumerate() {
-            let node = self.nodes[&node_id];
-            geometry.push(node.coord);
-            nodes.push(node.id);
-            if i == 0 {
-                source = node_id;
-            } else if node.uses > 1 {
-                result.push(Edge {
-                    id: format!("{}-{}", way.id.0, result.len()),
-                    osm_id: way.id,
-                    source,
-                    target: node_id,
-                    geometry,
-                    properties: way.properties,
-                    nodes,
-                });
-
-                source = node_id;
-                geometry = vec![node.coord];
-                nodes = vec![node.id]
-            }
-        }
-        result
+    let removed_ways_percentage = (removed_ways_count as f32 / self.ways.len() as f32) * 100.0;
+
+    // remove ways with missing nodes
+    for way_index in ways_to_remove.iter().rev() {
+        self.ways.remove(*way_index);
     }
 
-    fn read_ways(&mut self, file: std::fs::File) {
-        let mut pbf = osmpbfreader::OsmPbfReader::new(file);
-        for obj in pbf.iter().flatten() {
-            if let osmpbfreader::OsmObj::Way(way) = obj {
-                let mut skip = false;
-                let mut properties = EdgeProperties::default();
-                for (key, val) in way.tags.iter() {
-                    properties.update(key.to_string(), val.to_string());
-                    if self
-                        .forbidden
-                        .get(key.as_str())
-                        .map(|vals| vals.contains(val.as_str()) || vals.contains("*"))
-                        == Some(true)
-                    {
-                        skip = true;
-                    }
-                }
-                properties.normalize();
-                if properties.accessible() && !skip {
-                    for node in &way.nodes {
-                        self.nodes_to_keep.insert(*node);
-                    }
-                    self.ways.push(Way {
-                        id: way.id,
-                        nodes: way.nodes,
-                        properties,
-                    });
-                }
-            }
-        }
+    // print stats
+    for (way_index, node_id) in missing_node_stats {
+      println!("Way {} is missing node {}", way_index, node_id);
     }
+    println!("{} ways removed ({:.2}%) because of missing nodes", removed_ways_count, removed_ways_percentage);
```

Comment:
It would be good to know which nodes were removed from which ways.

Reply:
Currently, it is not missing nodes that are deleted, but all ways with at least one missing node.